### PR TITLE
[BinaryPlatforms] Change "shortest match" algorithm to "best match"

### LINKED ISSUE #49491: `Base.BinaryPlatforms.select_platform` broken on `master` and 1.9
Reproducer:
```julia
using Pkg, Base.BinaryPlatforms, Artifacts

mktempdir() do dir
    Pkg.activate(dir) do
        name = "Zlib_jll"
        spec = PackageSpec(
            ;
            name,
            uuid = "83775a58-1f1d-513f-b197-d71354ab007a",
            tree_hash = Base.SHA1("3c163199e21e5a0bab8881d6794496d02f28013d"),
            repo = Pkg.Types.GitRepo(
                ;
                rev="2a651eb0845ad1aeb4f65f17669216698cc216b2",
                source="https://github.com/JuliaBinaryWrappers/Zlib_jll.jl.git"
            ),
        )
        dep_path = Pkg.Operations.find_installed(name, spec.uuid, spec.tree_hash)
        artifacts_toml = joinpath(dep_path, "Artifacts.toml")
        platform = Platform("x86_64", "linux"; sanitize="memory")
        artifact_meta(name[1:end-4], artifacts_toml; platform)
    end
end
```
Expected output (and actual output on Julia v1.8):
```
Dict{String, Any} with 6 entries:
  "arch"          => "x86_64"
  "git-tree-sha1" => "d5fb5e2a1b66832e6c1a7234bc674e1e0e3eb0e9"
  "libc"          => "glibc"
  "download"      => Any[Dict{String, Any}("sha256"=>"13bffbc5a168fe4095553aec25c2d56102d8721330cb79b62f2359006432f748", "url"=>"https://github.com/JuliaBinaryWrappers/Zlib_jll.jl/releases/download/Zlib-v1.2.1…
  "os"            => "linux"
  "sanitize"      => "memory"
```
Actual output:
```
Dict{String, Any} with 5 entries:
  "arch"          => "x86_64"
  "git-tree-sha1" => "981cf7f0dcc3e5981e98dc47845f43ea8a07d5e0"
  "libc"          => "glibc"
  "download"      => Any[Dict{String, Any}("sha256"=>"d79c9a6ac68fc363ce55b881bcfb2667829172af7113ba187ddbb6dfef36a4a6", "url"=>"https://github.com/JuliaBinaryWrappers/Zlib_jll.jl/releases/download/Zlib-v1.2.1…
  "os"            => "linux"
```

Likely culprit: #48749.  CC: @staticfloat

### PR DESCRIPTION
**Pull request by staticfloat**
My assertion in the previous attempt to fix this issue was incorrect:

> We define a simpler match as one that has fewer tags overall.
> As these candidate matches have already been filtered to match the
> given platform, the only other tags that exist are ones that are in
> addition to the tags declared by the platform.  Hence, selecting the
> minimum in number of tags is equivalent to selecting the closest match.

This is demonstrably false, by my own test case:

```
platforms = Dict(
    Platform("x86_64", "linux") => "bad",
    Platform("x86_64", "linux"; sanitize="memory") => "good",
)
select_platform(platforms, Platform("x86_64", "linux"; sanitize="memory")) == "good"
```

In this case, because there exists a candidate that is _more general_ than the provided platform type, the shortest match is no longer the best match.

This PR performs a more rigorous matching that works more reliably in all cases.

### CODE REVIEW

**Comment by staticfloat** [comment]
This fixes https://github.com/JuliaLang/julia/issues/49491

**Comment by staticfloat** [comment]
@StefanKarpinski I'm totally open to the idea of using natural sort here; but such a sort function needs to be available pretty early on in the bootstrap process; do we have one?

File: test/binaryplatforms.jl [hunk 1]
```diff
@@ -330,8 +330,7 @@ end
     # Ambiguity test
     @test select_platform(platforms, P("aarch64", "linux")) == "linux3"
     @test select_platform(platforms, P("aarch64", "linux"; libgfortran_version=v"3")) == "linux3"
-    # This one may be surprising, but we still match `linux3`, and since linux3 is shorter, we choose it.
-    @test select_platform(platforms, P("aarch64", "linux"; libgfortran_version=v"3", libstdcxx_version=v"3.4.18")) === "linux3"
+    @test select_platform(platforms, P("aarch64", "linux"; libgfortran_version=v"3", libstdcxx_version=v"3.4.18")) === "linux5"
```
Comment (staticfloat) [inline test/binaryplatforms.jl:333]: As an aside, this "surprise" should have been my hint that my previous implementation was incorrect.

**Comment by giordano** [comment]
I think this looks better now, thanks!

**Comment by KristofferC** [comment]
Removing backport since original PR causing this was reverted on 1.9.

**Pull request merged**